src/modules/pipeline.py

```diff
@@ -1,10 +1,183 @@
+from __future__ import annotations
+import os
+import json
+from typing import Any, Union
 
+from torcheval.metrics.functional import multiclass_f1_score, multiclass_accuracy
+from torch.utils.data import DataLoader
 
+import src.modules.data_module as dm
+import src.modules.learner_module as lm
+import src.modules.evaluation_module as em
 
 
+class Pipeline:
+     
+    __slots__ = ['dataset', 'learner', 'settings', 'optimiser', 'loss_function',
+                 'train_loader', 'test_loader']
+    
+    def __init__(self, dataset: dm.ActiveDataset, learner: lm.ActiveLearner,
+                 optimiser, loss_function, settings: Union[dict, PipelineSettings]):
+        self.dataset = dataset
+        self.learner = learner
+        self.optimiser = optimiser
+        self.loss_function = loss_function
 
-class Pipeline():
+        self.test_loader = DataLoader(dataset.test_set, batch_size=128)
+        self.train_loader = DataLoader(dataset.labeled_set, batch_size=32, shuffle=True)
 
-    # TODO What parameters
+        if type(settings) == dict:
+            settings = PipelineSettings.from_dict(settings)
+        self.settings: PipelineSettings = settings
 
-    pass
\ No newline at end of file
+    def run(self) -> tuple[em.Session, dict[str, list[float]]]:
+        """
+        Runs the pipeline. Return a Session object as well as a dict of stats, which contains
+        loss history on train and test and accuracy history
+
+        Pseudocode:
+
+        perform initial learner fit
+        loop for n_queries
+              get_queries
+              label_queries
+              fit learner on updated train set
+              update session
+        return session
+        """
+        # initialise loss/accuracy history arrays
+        # initial learner fit
+        t_loss_history, v_loss_history = self.learner.fit(
+            self.train_loader, self.test_loader,
+            self.optimiser, self.loss_function,
+            epochs=self.settings.init_epochs,
+            early_stopping=True,
+        )
+        accuracy_history = [self._calculate_accuracy(self.test_loader).item()]
+
+        # initialise the session
+        session = em.Session(dataset=self.dataset, learner=self.learner)
+
+        for i in range(self.settings.n_queries):
+            print(f"Iteration: {i+1}")
+
+            # generate queries
+            unlabeled_loader = DataLoader(self.dataset.unlabeled_set, batch_size=128, shuffle=False)
+            queries, uncertainty = self.learner.generate_query(
+                unlabeled_loader,
+                criterion='entropy',
```

Comment:
Make this a parameter of PipelineSettings. Possible values: `confidence`, `margin`, `entropy`, `random`.